File: backend/src/main/java/com/evalorithm/service/TopicService.java
package com.evalorithm.service;

import com.evalorithm.dto.request.TopicRequest;
import com.evalorithm.dto.response.PageResponse;
import com.evalorithm.dto.response.TopicResponse;

import java.util.List;
import org.springframework.data.domain.Pageable;

public interface TopicService {

    PageResponse<TopicResponse> getAll(Pageable pageable);

    TopicResponse getById(Long id);

    List<TopicResponse> getByUnit(Long unitId);

    TopicResponse create(TopicRequest request);

    TopicResponse update(Long id, TopicRequest request);

    void delete(Long id);
}

File: frontend/src/services/topicService.ts
import api from './api';
import { Topic, TopicRequest, PageResponse, ApiResponse } from '../types';

const topicService = {
  getAll: async (
    page: number = 0,
    size: number = 10,
    sortBy: string = 'id',
    sortDir: string = 'asc'
  ): Promise<ApiResponse<PageResponse<Topic>>> => {
    const params = new URLSearchParams();
    params.append('page', page.toString());
    params.append('size', size.toString());
    params.append('sortBy', sortBy);
    params.append('sortDir', sortDir);
    const response = await api.get<ApiResponse<PageResponse<Topic>>>(
      `/topics?${params.toString()}`
    );
    return response.data;
  },

  getByUnit: async (unitId: number): Promise<ApiResponse<Topic[]>> => {
    const response = await api.get<ApiResponse<Topic[]>>(`/topics/unit/${unitId}`);
    return response.data;
  },

  create: async (data: TopicRequest): Promise<ApiResponse<Topic>> => {
    const response = await api.post<ApiResponse<Topic>>('/topics', data);
    return response.data;
  },

  update: async (id: number, data: TopicRequest): Promise<ApiResponse<Topic>> => {
    const response = await api.put<ApiResponse<Topic>>(`/topics/${id}`, data);
    return response.data;
  },

  delete: async (id: number): Promise<ApiResponse<void>> => {
    const response = await api.delete<ApiResponse<void>>(`/topics/${id}`);
    return response.data;
  },
};

export default topicService;
